File: articles/data-measurement-reproducibility-life-sciences/r/qc_summary.R
# QC summary by batch.

data_path <- file.path("data", "measurements.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "measurements.csv")
}

measurements <- read.csv(data_path)

qc_counts <- aggregate(
  sample_id ~ batch_id + qc_flag,
  data = measurements,
  FUN = length
)

names(qc_counts)[names(qc_counts) == "sample_id"] <- "n_records"

batch_totals <- aggregate(
  sample_id ~ batch_id,
  data = measurements,
  FUN = length
)

names(batch_totals)[names(batch_totals) == "sample_id"] <- "batch_total"

summary_df <- merge(qc_counts, batch_totals, by = "batch_id")
summary_df$flag_fraction <- summary_df$n_records / summary_df$batch_total

print(summary_df[order(summary_df$batch_id, summary_df$qc_flag), ])

File: articles/r-biological-data-analysis-visualization/python/qc_summary.py
"""
Quality-control summary for biological measurements.

Run:
    python python/qc_summary.py
"""

from pathlib import Path

import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]


def main() -> None:
    data = pd.read_csv(ARTICLE_DIR / "data" / "measurements.csv")

    summary = (
        data.groupby(["batch_id", "qc_flag"])
        .size()
        .reset_index(name="n_records")
        .sort_values(["batch_id", "qc_flag"])
    )

    batch_totals = data.groupby("batch_id").size().reset_index(name="batch_total")
    summary = summary.merge(batch_totals, on="batch_id")
    summary["flag_fraction"] = summary["n_records"] / summary["batch_total"]

    print(summary.round(5).to_string(index=False))


if __name__ == "__main__":
    main()
